src/chttp.cpp

```diff
@@ -12,28 +12,36 @@ std::string getUserAgent() {
 	return (std::string) "curl/" + info->version + " gmod-chttp/" + CHTTP_VERSION;
 }
 
-void runFailedHandler(Lua::ILuaBase *LUA, int handler, const std::string& reason) {
-	if (!handler)
+void runFailedHandler(Lua::ILuaBase *LUA, int successHandler, int failHandler, const std::string& reason) {
+	if(!failHandler) {
+		goto failedCleanup;
 		return;
+	}
 
 	// Push fail handler to stack and free our ref
-	LUA->ReferencePush(handler);
-	LUA->ReferenceFree(handler);
+	LUA->ReferencePush(failHandler);
 
 	// Push the argument
 	LUA->PushString(reason.c_str());
 
 	// Call the fail handler with one argument
 	LUA->Call(1, 0);
+
+	LUA->ReferenceFree(failHandler);
+
+failedCleanup:
+	if(successHandler)
+		LUA->ReferenceFree(successHandler);
```

Comment:
What about doing this right at the beginning? That would keep the diff more minimal, and we don't need the successHandler anyways while handling failures. The same would apply for `runSuccessHandler()`.